subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy

```diff
@@ -1431,4 +1431,65 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
         1 * action.execute(a)
         0 * action.execute(_)
     }
+
+    protected def getInvalidCallFromLazyConfiguration() {
+        return [
+            ["add(T)"               , CallAddFactory.AsAction],
+            ["add(T)"               , CallAddFactory.AsClosure],
+            ["addLater(Provider)"   , CallAddLaterFactory.AsAction],
+            ["addLater(Provider)"   , CallAddLaterFactory.AsClosure],
+            ["addAllLater(Provider)", CallAddAllLaterFactory.AsAction],
+            ["addAllLater(Provider)", CallAddAllLaterFactory.AsClosure],
+            ["addAll(Collection)"   , CallAddAllFactory.AsAction],
+            ["addAll(Collection)"   , CallAddAllFactory.AsClosure],
+            ["clear()"              , CallClearFactory.AsAction],
+            ["clear()"              , CallClearFactory.AsClosure],
+            ["remove(Object)"       , CallRemoveFactory.AsAction],
+            ["remove(Object)"       , CallRemoveFactory.AsClosure],
+            ["removeAll(Collection)", CallRemoveAllFactory.AsAction],
+            ["removeAll(Collection)", CallRemoveAllFactory.AsClosure],
+            ["retainAll(Collection)", CallRetainAllFactory.AsAction],
+            ["retainAll(Collection)", CallRetainAllFactory.AsClosure],
+            ["iterator().remove()"  , CallRemoveOnIteratorFactory.AsAction],
+            ["iterator().remove()"  , CallRemoveOnIteratorFactory.AsClosure],
+        ]
+    }
+
+    @Unroll
+    def "disallow mutating when configureEach(#factoryClass.configurationType.simpleName) calls #description"() {
```

Comment:
We also need some coverage that an action can call a query method on the container. This includes the providers that the container creates. e.g:

```
def a = tasks.register("a")
tasks.register("b") {
    a.get()
   tasks.getByName("a")
}
```